File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/ChatDetail/View/Message/MSVideoMessageCell.m
//
//  MSVideoMessageCell.m
//  BlackFireIM
//
//  Created by benny wang on 2021/4/2.
//

#import "MSVideoMessageCell.h"
#import "MSIMSDK-UIKit.h"
#import <SDWebImage.h>


@implementation MSVideoMessageCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        _thumb = [[UIImageView alloc] init];
        _thumb.layer.cornerRadius = 5.0;
        [_thumb.layer setMasksToBounds:YES];
        _thumb.contentMode = UIViewContentModeScaleAspectFit;
        _thumb.backgroundColor = [UIColor whiteColor];
        [self.container addSubview:_thumb];
        _thumb.frame = self.container.bounds;
        _thumb.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;

        CGSize playSize = TVideoMessageCell_Play_Size;
        _playIcon = [[UIImageView alloc]initWithFrame:CGRectMake(0, 0, playSize.width, playSize.height)];
        _playIcon.contentMode = UIViewContentModeScaleAspectFit;
        _playIcon.image = [UIImage imageNamed:TUIKitResource(@"play_normal")];
        [self.container addSubview:_playIcon];
        
        _durationL = [[UILabel alloc]init];
        _durationL.textColor = [UIColor whiteColor];
        _durationL.font = [UIFont systemFontOfSize:12];
        [self.container addSubview:_durationL];
        
        _progressL = [[UILabel alloc] init];
        _progressL.textColor = [UIColor whiteColor];
        _progressL.font = [UIFont systemFontOfSize:15];
        _progressL.textAlignment = NSTextAlignmentCenter;
        _progressL.layer.cornerRadius = 5.0;
        _progressL.hidden = YES;
        _progressL.backgroundColor = TImageMessageCell_Progress_Color;
        [_progressL.layer setMasksToBounds:YES];
        [self.container addSubview:_progressL];
        _progressL.frame = self.container.bounds;
        _progressL.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    }
    return self;
}

- (MSVideoMessageCellData *)videoData
{
    return (MSVideoMessageCellData *)self.messageData;
}

- (void)fillWithData:(MSVideoMessageCellData *)data
{
    [super fillWithData:data];
    self.durationL.text = [NSString stringWithFormat:@"%02zd:%02zd",data.message.videoElem.duration/60,data.message.videoElem.duration%60];
    [self.durationL sizeToFit];
    NSInteger progress = data.message.videoElem.progress*100;
    self.progressL.text = [NSString stringWithFormat:@"%zd%%",progress];
    [self.progressL setHidden:!(progress > 0 && progress < 100)];
    self.thumb.image = nil;
    if (data.message.videoElem.coverImage) {
        self.thumb.image = data.message.videoElem.coverImage;
    }else if ([[NSFileManager defaultManager]fileExistsAtPath:data.message.videoElem.coverPath]) {
        UIImage *image = [UIImage imageWithContentsOfFile:data.message.videoElem.coverPath];
        self.thumb.image = image;
        data.message.videoElem.coverImage = image;
    }else {
        NSString *thumbUrl = [NSString stringWithFormat:@"%@?imageMogr2/thumbnail/300x/interlace/0",data.message.videoElem.coverUrl];
        [self.thumb sd_setImageWithURL:[NSURL URLWithString:thumbUrl] placeholderImage:[UIImage imageNamed:TUIKitResource(@"placeholder_delete")] options:SDWebImageDelayPlaceholder];
    }
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    self.playIcon.center = CGPointMake(self.container.width*0.5, self.container.height*0.5);
    self.durationL.maxY = self.container.height-TVideoMessageCell_Margin_3;
    self.durationL.maxX = self.container.width-TVideoMessageCell_Margin_3;
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/ChatDetail/View/Message/MSVideoMessageCell.swift
//
//  MSVideoMessageCell.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/10.
//

import UIKit
import Kingfisher



open class MSVideoMessageCell: MSMessageCell {

    public var thumb: UIImageView!
    
    public var durationL: UILabel!
    
    public var playIcon: UIImageView!
    
    public var progressL: UILabel!
    
    public var videoData: MSVideoMessageCellData?
    
    public override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        thumb = UIImageView()
        thumb.layer.cornerRadius = 5
        thumb.layer.masksToBounds = true
        thumb.contentMode = .scaleAspectFit
        thumb.backgroundColor = .white
        container.addSubview(thumb)
        thumb.frame = container.bounds
        thumb.autoresizingMask = [.flexibleWidth,.flexibleHeight]
        
        let playSize = MSMcros.TVideoMessageCell_Play_Size
        playIcon = UIImageView(frame: CGRect(x: 0, y: 0, width: playSize.width, height: playSize.height))
        playIcon.contentMode = .scaleAspectFit
        playIcon.image = UIImage.bf_imageNamed(name: "play_normal")
        container.addSubview(playIcon)
        
        durationL = UILabel()
        durationL.textColor = .white
        durationL.font = .systemFont(ofSize: 12)
        container.addSubview(durationL)
        
        progressL = UILabel()
        progressL.textColor = .white
        progressL.font = .systemFont(ofSize: 15)
        progressL.textAlignment = .center
        progressL.layer.cornerRadius = 5
        progressL.isHidden = true
        progressL.backgroundColor = MSMcros.TImageMessageCell_Progress_Color
        progressL.layer.masksToBounds = true
        container.addSubview(progressL)
        progressL.frame = container.bounds
        progressL.autoresizingMask = [.flexibleWidth,.flexibleHeight]
    }
    
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    public override func fillWithData(data: MSMessageCellData) {
        super.fillWithData(data: data)
        if let videoData = data as? MSVideoMessageCellData {
            self.videoData = videoData
            self.durationL.text = String(format: "%02zd:%02zd", videoData.videoElem.duration / 60, videoData.videoElem.duration % 60)
            self.durationL.sizeToFit()
            let progress = videoData.videoElem.progress * 100
            self.progressL.text = String.init(format: "%.1f%%", videoData.videoElem.progress * 100)
            self.progressL.isHidden = !(progress > 0 && progress < 100)
            thumb.image = nil
            if videoData.videoElem.coverImage != nil {
                thumb.image = videoData.videoElem.coverImage
            }else if videoData.videoElem.coverPath != nil && FileManager.default.fileExists(atPath: videoData.videoElem.coverPath!) {
                let image = UIImage(contentsOfFile: videoData.videoElem.coverPath!)!
                thumb.image = image
                videoData.videoElem.coverImage = image
            }else if videoData.videoElem.coverUrl != nil {
                let thumbUrl = videoData.videoElem.coverUrl! + "?imageMogr2/thumbnail/300x/interlace/0"
                thumb.kf.setImage(with: URL(string: thumbUrl), options: [KingfisherOptionsInfoItem.onFailureImage(UIImage.bf_imageNamed(name: "placeholder_delete"))])
            }
        }
    }

    open override func layoutSubviews() {
        super.layoutSubviews()
        playIcon.center = CGPoint(x: container.width * 0.5, y: container.height * 0.5)
        durationL.bottom = container.height - MSMcros.TVideoMessageCell_Margin_3
        durationL.right = container.width  - MSMcros.TVideoMessageCell_Margin_3
    }
}
